### DISCUSSION TITLE
Calling polybar-msg in custom/ipc hook fails.

### DISCUSSION THREAD

**Discussion by llc0930** [polybar/polybar#2920] [Q&A] [answered]
`polybar -v`
polybar 3.6.3

Features: +alsa +curl +i3 +mpd +network(libnl) +pulseaudio +xkeyboard
```
[bar/root]
.
. (skip)
.
enable-ipc = true
.
. (skip)
.
[bar/center]
inherit = bar/root
.
. (skip)
.
modules-center = demo demo-2 xwindow
.
. (skip)
.
[module/demo]
type = custom/ipc
hook-0 = echo "%pid%"
hook-1 = polybar-msg action demo-2 send %pid%
initial = 1

[module/demo-2]
type = custom/ipc
hook-0 = echo "wait"
initial = 1
```
`polybar-msg action demo hook 1`
polybar-msg stuck in do_epoll_wait.
Any ideas?

**Comment by patrick96** [comment, marked as answer]
That's an interesting problem. The ipc module waits for `hook-1` to finish executing while the command in `hook-1` waits for polybar to respond. Polybar can't respond because it is waiting for the hook to finish. Classical deadlock.

What is your use-case for doing this?

The only solution for now is to not call `polybar-msg` from an IPC hook. Or if you do, make sure to completely detach it with something like `setsid`.

**Reply by llc0930** [reply]
`nohup polybar-msg action demo-2 send %pid% &`
`(polybar-msg action demo-2 send %pid% &)`
![2023-02-17-01:21:08-r](https://user-images.githubusercontent.com/14966910/219440505-fccd9bc3-7876-42ca-9434-457c877524bb.png)

`setsid polybar-msg action demo-2 send %pid%`
![2023-02-17-01:24:57-r](https://user-images.githubusercontent.com/14966910/219441368-a4276f5f-6d1b-409e-93e7-0436378b307e.png)

I'm not sure, is there any other way?

When the workspace is empty, I want to make barB obtain the pid of barA through polybar-msg without recording the pid of barA before, and then notify barB to  use polybar-msg hook hide barA.
Reverse operation when workspace not empty, let barA show.

**Reply by llc0930** [reply]
![2023-02-17-01:44:42-r](https://user-images.githubusercontent.com/14966910/219446192-0253801a-2da9-483f-8aa0-f3f5da383af3.png)
![2023-02-17-01:45:15-r](https://user-images.githubusercontent.com/14966910/219446203-6810cc41-aa46-447f-889c-32bfbf0c65a2.png)
![2023-02-17-01:45:33-r](https://user-images.githubusercontent.com/14966910/219446220-f5ec2510-5588-4f3f-84f4-d2547ed63576.png)
When the bar is different, the pid send by polybar-msg can be received, but the polybar-msg is still stuck there.

**Reply by llc0930** [reply]
@patrick96, I get a zombie process by
`hook-0 = nohup sh -c 'sleep 5 && polybar-msg -p %pid% cmd hide' &`
Is this normal?
![2023-02-17-23:03:46](https://user-images.githubusercontent.com/14966910/219694064-6edcd8de-000c-494b-8466-c8bf8ea87788.png)
![2023-02-17-23:06:41](https://user-images.githubusercontent.com/14966910/219694092-0921c097-49d2-46d0-a367-7163826d556c.png)
![2023-02-17-23:18:39-r](https://user-images.githubusercontent.com/14966910/219694114-1b62e518-d0ff-463f-b63d-1f25edfdbe88.png)

**Comment by llc0930** [comment]
```
[bar/root]
.
. (skip)
.
enable-ipc = true
.
. (skip)
.
[bar/center]
inherit = bar/root
.
. (skip)
.
modules-center = demo xwindow
.
. (skip)
.
[module/demo]
type = custom/ipc
hook-0 = alacritty -e sh -c 'nohup polybar-msg -p %pid% cmd hide & sleep 0.1'
hook-1 = alacritty -e sh -c 'nohup polybar-msg -p %pid% cmd show & sleep 0.1'
hidden = true
initial = 0
```
Finally got it work temporarily.
Maybe set a title for the terminal emulator so that wm can drop it to another workspace.

- failed group1
```
hook-0 = alacritty -e 'polybar-msg -p %pid% cmd hide'
hook-1 = alacritty -e 'polybar-msg -p %pid% cmd show'
;hidden = true
```
The ipc module show Error: "Event loop terminated with code: 1"
![2023-02-17-05:03:49](https://user-images.githubusercontent.com/14966910/219487205-f411f451-cbf3-4890-aa18-d3f47bfd77e1.png)
- failed group2
```
hook-0 = alacritty -e sh -c 'polybar-msg -p %pid% cmd hide'
hook-1 = alacritty -e sh -c 'polybar-msg -p %pid% cmd show'
```
After ctrl+c leaves the terminal emulator, the bar is hidden and polybar-msg exits, but the bar no longer responds.
![2023-02-17-04:51:46](https://user-images.githubusercontent.com/14966910/219485583-3ab985be-6db2-4931-ae5c-b8d8ae03ea37.png)
![2023-02-17-04:52:40](https://user-images.githubusercontent.com/14966910/219485596-c625b0ef-25bf-47bd-8b3c-33d73aaedce6.png)
- failed group3
```
hook-0 = alacritty -e sh -c 'nohup polybar-msg -p %pid% cmd hide & '
hook-1 = alacritty -e sh -c 'nohup polybar-msg -p %pid% cmd show & '
```
Compared to the successful group, nothing happened, but bar didn't lose response or report errors.

**Reply by llc0930** [reply]
Finally, I found that it can be easily achieved by using the pgrep command.
`polybar-msg -p $(pgrep -f "polybar center") cmd hide`
`polybar-msg -p $(pgrep -f "polybar center") cmd show`
